internal/storage/writer/pubsub/pubsub.go

```diff
@@ -0,0 +1,139 @@
+package pubsub
+
+import (
+	"context"
+	"runtime"
+	"time"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/grab/async"
+	"github.com/kelindar/talaria/internal/encoding/block"
+	"github.com/kelindar/talaria/internal/encoding/key"
+	"github.com/kelindar/talaria/internal/monitor"
+	"github.com/kelindar/talaria/internal/monitor/errors"
+	script "github.com/kelindar/talaria/internal/scripting"
+	"github.com/kelindar/talaria/internal/storage/writer/base"
+	"google.golang.org/api/option"
+)
+
+// Writer to write and streaming to PubSub
+type Writer struct {
+	*base.Writer
+	client  *pubsub.Client
+	topic   *pubsub.Topic
+	monitor monitor.Monitor
+	context context.Context
+	buffer  chan []byte
+	queue   chan async.Task
+	task    async.Task
+}
+
+// New creates a new writer
+func New(project, topic, encoding, filter string, loader *script.Loader, monitor monitor.Monitor, opts ...option.ClientOption) (*Writer, error) {
+	ctx := context.Background()
+	client, err := pubsub.NewClient(ctx, project, opts...)
+
+	if err != nil {
+		return nil, errors.Newf("pubsub: %v", err)
+	}
+
+	// Load encoder
+	encoderWriter, err := base.New(filter, encoding, loader)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check if topic exists
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+	topicRef := client.Topic(topic)
+	ok, err := topicRef.Exists(ctx)
+
+	if err != nil {
+		return nil, errors.Newf("pubsub: %v", err)
+	}
+	if !ok {
+		return nil, errors.New("pubsub: topic does not exist")
+	}
+
+	w := &Writer{
+		topic:   topicRef,
+		client:  client,
+		Writer:  encoderWriter,
+		monitor: monitor,
+		context: ctx,
+		buffer:  make(chan []byte, 65000),
+		queue:   make(chan async.Task),
+	}
+	w.Process = w.process
+
+	return w, nil
+}
+
+// Write writes the data to the sink.
+func (w *Writer) Write(key.Key, []byte) error {
+	return nil // Noop
+}
+
+// Stream publishes data into PubSub topics
+func (w *Writer) Stream(row block.Row) error {
+	message, err := w.Writer.Encode(row)
+	if err != nil {
+		return err
+	}
+
+	// If message is filtered out, return nil
+	if message == nil {
+		return nil
+	}
+
+	select {
+	case w.buffer <- message:
+	default:
+		return errors.New("pubsub: buffer is full")
+	}
+	return nil
+}
+
+// process will read from buffer, encode the row, and publish to PubSub
+func (w *Writer) process(parent context.Context) error {
+	w.task = async.Consume(parent, runtime.NumCPU()*8, w.queue)
+	for message := range w.buffer {
+		select {
```

Comment:
There's a data race on the `message` I think:
```
WARNING: DATA RACE
Write at 0x00c00051a040 by goroutine 35:
  github.com/kelindar/talaria/internal/storage/writer/pubsub.(*Writer).process()
      /home/runner/work/talaria/talaria/internal/storage/writer/pubsub/pubsub.go:101 +0x21c
  github.com/kelindar/talaria/internal/storage/writer/pubsub.(*Writer).process-fm()
      /home/runner/work/talaria/talaria/internal/storage/writer/pubsub/pubsub.go:99 +0x55
  github.com/kelindar/talaria/internal/storage/writer/base.(*Writer).Run.func1()
      /home/runner/work/talaria/talaria/internal/storage/writer/base/base.go:77 +0x71
  github.com/grab/async.(*task).run.func1()
      /home/runner/go/pkg/mod/github.com/grab/async@v0.0.5/task.go:139 +0x63

Previous read at 0x00c00051a040 by goroutine 40:
  github.com/kelindar/talaria/internal/storage/writer/pubsub.(*Writer).process.func1()
      /home/runner/work/talaria/talaria/internal/storage/writer/pubsub/pubsub.go:116 +0x192
  github.com/grab/async.(*task).run.func1()
      /home/runner/go/pkg/mod/github.com/grab/async@v0.0.5/task.go:139 +0x63

```

Reply:
I think we need to make a copy of the message before pushing it into the queue. Something like this:
```
encoded := message // copy the message
w.queue <- async.NewTask(func(ctx context.Context) (interface{}, error) {
   ...
   return nil, w.processResult(result, encoded) // use encoded message to avoid data race
})
```